### PR TITLE
Add a Pip package with tools for master notebook authoring

### PR DESCRIPTION
**Pull request by salikh**

### CODE REVIEW

File: python/prog_edu_assistant_tools/prog_edu_assistant_tools/summary_test_result.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
+"""A package providing SummaryTestResult.
+
+This pacakge provides SummaryTestResult, which is an extensions
```
Comment (victorkwan) [inline python/prog_edu_assistant_tools/prog_edu_assistant_tools/summary_test_result.py]:
```suggestion
This package provides SummaryTestResult, which is an extension
```

File: python/prog_edu_assistant_tools/prog_edu_assistant_tools/magics.py [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+    """Non-magic version of the autotest.
+
+    This function can be used as
+
+        result, log = autotest(MyTestCase)
+
+    """
+    suite = unittest.TestLoader().loadTestsFromTestCase(testClass)
+    errors = io.StringIO()
+    result = unittest.TextTestRunner(verbosity=4,
```
Comment (victorkwan) [inline python/prog_edu_assistant_tools/prog_edu_assistant_tools/magics.py]: Just out of curiosity, what does this level of verbosity give us?

File: python/prog_edu_assistant_tools/prog_edu_assistant_tools/magics.py [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+                                     resultclass=SummaryTestResult).run(suite)
+    return result, errors.getvalue()
+
+
+def report(template, **kwargs):
+    """Non-magic version of the report.
+
+    This function can be used as
+
+        report(template, submission_source.source, results)
```
Comment (victorkwan) [inline python/prog_edu_assistant_tools/prog_edu_assistant_tools/magics.py]: Should this document that the kwargs are passed into template.render?

File: python/prog_edu_assistant_tools/prog_edu_assistant_tools/magics.py [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+        'my cell magic'
+        return line, cell
+
+    @cell_magic
+    def submission(self, line, cell):
+        """Registers a submission_source and submission, if the code can run.
+
+        This magic is useful for auto-testing (testing autograder unit tests on
+        incorrect inputs)
+    """
```
Comment (victorkwan) [inline python/prog_edu_assistant_tools/prog_edu_assistant_tools/magics.py]:
```suggestion
        """
```

File: python/prog_edu_assistant_tools/prog_edu_assistant_tools/magics.py [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+# The class MUST call this class decorator at creation time
+@magics_class
+class MyMagics(Magics):
+    @line_magic
+    def autotest(self, line):
+        """Run the unit tests inline
+
+        Returns the result object. result.results is a summary dictionary of
+        outcomes.
+        result.errors
```
Comment (victorkwan) [inline python/prog_edu_assistant_tools/prog_edu_assistant_tools/magics.py]: This seems to stop mid-sentence 🧐

File: python/prog_edu_assistant_tools/prog_edu_assistant_tools/magics.py [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+    def solution(self, line, cell):
+        """Registers solution and evaluates it.
+
+        Also removes the PROMPT block and %%solution from the solution source.
+
+        The difference from %%submission is that the solution is copied into the
+        top context,
+        making it possible to refer to the functions and variables in subsequent
+        notebook cells.
+    """
```
Comment (victorkwan) [inline python/prog_edu_assistant_tools/prog_edu_assistant_tools/magics.py]:
```suggestion
        """
```

File: python/prog_edu_assistant_tools/prog_edu_assistant_tools/magics.py [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+        """Registers a template for report generation.
+
+        Use {{results['TestClassName.test_method']}} to extract specific
+        outcomes and be prepared
+        for the keys to be absent in the results map, if the test could not be
+        run at all (e.g. because
+        of syntax error in the submission).
+        """
+        name = line
+        if line == '':
```
Comment (victorkwan) [inline python/prog_edu_assistant_tools/prog_edu_assistant_tools/magics.py]: Can we document in what situation `line` may be empty?

File: python/prog_edu_assistant_tools/prog_edu_assistant_tools/magics.py (hunk 2 above)
Comment (salikh) [reply python/prog_edu_assistant_tools/prog_edu_assistant_tools/magics.py]: Okay, maybe level 4 is not necessary, but without explicitly verbosity of at least 2 the snippets like these are not shown for passing tests:
test_has_hello (HelloOutputTest.HelloOutputTest) ... ok
test_not_empty (HelloOutputTest.HelloOutputTest) ... ok
test_output (HelloOutputTest.HelloOutputTest) ... ok
Compare (verbosity 1):
image
to (verbosity 2):
image

File: python/prog_edu_assistant_tools/prog_edu_assistant_tools/magics.py (hunk 3 above)
Comment (salikh) [reply python/prog_edu_assistant_tools/prog_edu_assistant_tools/magics.py]: Yep, good idea. Done.

File: python/prog_edu_assistant_tools/prog_edu_assistant_tools/magics.py (hunk 7 above)
Comment (salikh) [reply python/prog_edu_assistant_tools/prog_edu_assistant_tools/magics.py]: Done.

Comment (salikh) [reply python/prog_edu_assistant_tools/prog_edu_assistant_tools/magics.py]: Done.

**Pull request merged**